src/main/scala/org/ergoplatform/nodeView/wallet/TrackedBoxSerializer.scala

```diff
@@ -0,0 +1,214 @@
+package org.ergoplatform.nodeView.wallet
+
+import org.ergoplatform.ErgoBox
+import org.ergoplatform.modifiers.mempool.{ErgoBoxSerializer, ErgoTransaction}
+import org.ergoplatform.nodeView.wallet.BoxCertainty.{Certain, Uncertain}
+import org.ergoplatform.nodeView.wallet.OnchainStatus.{Offchain, Onchain}
+import org.ergoplatform.nodeView.wallet.SpendingStatus.{Spent, Unspent}
+import org.ergoplatform.nodeView.wallet.TrackedBoxSerializer.TransactionLookup
+import org.ergoplatform.settings.Constants.ModifierIdSize
+import scorex.core.serialization.Serializer
+import scorex.core.utils.ScorexEncoding
+import scorex.core.validation.ModifierValidator
+import scorex.core.{ModifierId, bytesToId, idToBytes}
+import sigmastate.utils.{ByteReader, ByteWriter}
+
+import scala.util.{Failure, Try}
+
+class TrackedBoxSerializer(txLookup: TransactionLookup) extends TypedBoxSerializer[TrackedBox] {
+
+  def write(trackedBox: TrackedBox, w: ByteWriter): Unit = {
+    trackedBox match {
+      case b: UnspentOffchainBox => new UnspentOffchainBoxSerializer(txLookup).write(b, w)
+      case b: UnspentOnchainBox => new UnspentOnchainBoxSerializer(txLookup).write(b, w)
+      case b: SpentOffchainBox => new SpentOffchainBoxSerializer(txLookup).write(b, w)
+      case b: SpentOnchainBox => new SpentOnchainBoxSerializer(txLookup).write(b, w)
+    }
+  }
+
+  override def read(r: ByteReader): Try[TrackedBox] = {
+    val pos = r.position
+    val (_, spendingStatus, onchainStatus) = readHeaderBits(r)
+    r.position = pos
+
+    (spendingStatus, onchainStatus) match {
+      case (Unspent, Offchain) => new UnspentOffchainBoxSerializer(txLookup).read(r)
+      case (Unspent, Onchain) => new UnspentOnchainBoxSerializer(txLookup).read(r)
+      case (Spent, Offchain) => new SpentOffchainBoxSerializer(txLookup).read(r)
+      case (Spent, Onchain) => new SpentOnchainBoxSerializer(txLookup).read(r)
+    }
+  }
+
+}
+
+object TrackedBoxSerializer {
+  type TransactionLookup = ModifierId => Option[ErgoTransaction]
+}
+
+class UnspentOffchainBoxSerializer(txLookup: TransactionLookup) extends TypedBoxSerializer[UnspentOffchainBox] {
```

Comment:
Do we need this 4 separate case classes (both for serializers and boxes)? I think single class `TypedBoxSerializer` with parameters `(spendingStatus, onchainStatus)` would be enough and will reduce code size.
These classes implementations are very similar, read/write fields are almost copy-paste of each other

Reply:
Using one type for all box types reduces code size but also reduces type safety when using boxes and lacks code clarity when creating them. 

I'm about to finish code rearrangement for this